File: src/claude_stats/lines.py
"""Lines of code analytics — written, edited, and removed."""

import json
import sys
from collections import defaultdict

from claude_stats.common import (
    parse_dates, find_sessions, extract_project, shorten_path, get_ext,
    fmt, bar, print_header,
)


def count_lines(s: str) -> int:
    return len(s.splitlines()) if s else 0


def run(args: list[str] | None = None):
    if args is None:
        args = sys.argv[1:]
    dates, label = parse_dates(args)
    session_files = find_sessions(dates)

    if not session_files:
        print(f"\n  No sessions found for {label}")
        sys.exit(0)

    total_writes = 0
    total_edits = 0
    total_written = 0
    total_added = 0
    total_removed = 0
    total_files_created = 0
    total_files_edited = 0

    ext_stats = defaultdict(lambda: {"written": 0, "added": 0, "removed": 0, "files": set()})
    proj_stats = defaultdict(lambda: {"written": 0, "added": 0, "removed": 0, "writes": 0, "edits": 0, "files": set()})
    file_stats = defaultdict(lambda: {"written": 0, "added": 0, "removed": 0, "ops": 0})
    all_files = set()

    for sf in session_files:
        project = extract_project(sf)
        try:
            for line in open(sf):
                line = line.strip()
                if not line:
                    continue
                try:
                    obj = json.loads(line)
                except json.JSONDecodeError:
                    continue

                if obj.get("type") != "assistant":
                    continue
                msg = obj.get("message", {})
                if not isinstance(msg, dict):
                    continue

                for block in msg.get("content", []):
                    if not isinstance(block, dict) or block.get("type") != "tool_use":
                        continue

                    name = block.get("name")
                    inp = block.get("input", {})

                    if name == "Write":
                        fp = inp.get("file_path", "")
                        content = inp.get("content", "")
                        lines = count_lines(content)
                        total_writes += 1
                        total_written += lines
                        total_files_created += 1

                        ext = get_ext(fp)
                        ext_stats[ext]["written"] += lines
                        ext_stats[ext]["files"].add(fp)
                        proj_stats[project]["written"] += lines
                        proj_stats[project]["writes"] += 1
                        proj_stats[project]["files"].add(fp)

                        short = shorten_path(fp)
                        file_stats[short]["written"] += lines
                        file_stats[short]["ops"] += 1
                        all_files.add(fp)

                    elif name == "Edit":
                        fp = inp.get("file_path", "")
                        old = inp.get("old_string", "")
                        new = inp.get("new_string", "")
                        old_lines = count_lines(old)
                        new_lines = count_lines(new)
                        total_edits += 1
                        total_added += new_lines
                        total_removed += old_lines
                        total_files_edited += 1

                        ext = get_ext(fp)
                        ext_stats[ext]["added"] += new_lines
                        ext_stats[ext]["removed"] += old_lines
                        ext_stats[ext]["files"].add(fp)
                        proj_stats[project]["added"] += new_lines
                        proj_stats[project]["removed"] += old_lines
                        proj_stats[project]["edits"] += 1
                        proj_stats[project]["files"].add(fp)

                        short = shorten_path(fp)
                        file_stats[short]["added"] += new_lines
                        file_stats[short]["removed"] += old_lines
                        file_stats[short]["ops"] += 1
                        all_files.add(fp)
        except Exception:
            pass

    total_ops = total_writes + total_edits
    if total_ops == 0:
        print(f"\n  No Write/Edit operations found for {label}")
        sys.exit(0)

    net = total_written + total_added - total_removed

    # ── Summary
    print_header(f"📝  CLAUDE CODE LINES OF CODE — {label}")
    print(f"""
  Operations:    {total_writes} writes + {total_edits} edits = {total_ops} total
  Files touched: {len(all_files)}

  ┌──────────────────────────────────────┐
  │  Lines written (new files):  {fmt(total_written):>8} │
  │  Lines added (edits):        {fmt(total_added):>8} │
  │  Lines removed (edits):      {fmt(total_removed):>8} │
  │  ──────────────────────────────────  │
  │  NET LINES:                  {fmt(net):>8} │
  │  TOTAL THROUGHPUT:           {fmt(total_written + total_added + total_removed):>8} │
  └──────────────────────────────────────┘""")

    # ── By Extension
    print_header("📄  BY FILE EXTENSION", "─")
    sorted_ext = sorted(ext_stats.items(), key=lambda x: x[1]["written"] + x[1]["added"], reverse=True)[:15]
    max_ext = max((e["written"] + e["added"]) for _, e in sorted_ext) if sorted_ext else 1

    print(f"\n  {'Ext':<10} {'Written':>8} {'Added':>8} {'Removed':>8} {'Net':>8} {'Files':>5}  {'':>12}")
    print(f"  {'─' * 10} {'─' * 8} {'─' * 8} {'─' * 8} {'─' * 8} {'─' * 5}  {'─' * 12}")

    for ext, s in sorted_ext:
        ext_net = s["written"] + s["added"] - s["removed"]
        total_ext = s["written"] + s["added"]
        sign = "+" if ext_net >= 0 else ""
        print(f"  {ext:<10} {s['written']:>8} {s['added']:>8} {s['removed']:>8} {sign}{ext_net:>7} {len(s['files']):>5}  {bar(total_ext, max_ext, 12)}")

    # ── By Project
    print_header("📁  BY PROJECT", "─")
    sorted_proj = sorted(proj_stats.items(), key=lambda x: x[1]["written"] + x[1]["added"], reverse=True)
    max_proj = max((p["written"] + p["added"]) for _, p in sorted_proj) if sorted_proj else 1

    print(f"\n  {'Project':<40} {'Written':>8} {'Added':>8} {'Removed':>8} {'Net':>7}  {'':>10}")
    print(f"  {'─' * 40} {'─' * 8} {'─' * 8} {'─' * 8} {'─' * 7}  {'─' * 10}")

    for proj, s in sorted_proj:
        proj_net = s["written"] + s["added"] - s["removed"]
        total_proj = s["written"] + s["added"]
        name = proj[:38] if len(proj) > 38 else proj
        sign = "+" if proj_net >= 0 else ""
        print(f"  {name:<40} {s['written']:>8} {s['added']:>8} {s['removed']:>8} {sign}{proj_net:>6}  {bar(total_proj, max_proj, 10)}")

    # ── Top Files (most touched)
    print_header("🔥  TOP FILES (most operations)", "─")
    sorted_files = sorted(file_stats.items(), key=lambda x: x[1]["ops"], reverse=True)[:20]

    print(f"\n  {'Ops':>4} {'Written':>8} {'+Lines':>7} {'-Lines':>7}  File")
    print(f"  {'─' * 4} {'─' * 8} {'─' * 7} {'─' * 7}  {'─' * 45}")

    for fp, s in sorted_files:
        display = fp if len(fp) <= 50 else "..." + fp[-(47):]
        print(f"  {s['ops']:>4} {s['written']:>8} {s['added']:>7} {s['removed']:>7}  {display}")

    # ── Top Files (most lines)
    print_header("📏  TOP FILES (most lines written/added)", "─")
    sorted_by_lines = sorted(file_stats.items(), key=lambda x: x[1]["written"] + x[1]["added"], reverse=True)[:15]
    max_lines = (sorted_by_lines[0][1]["written"] + sorted_by_lines[0][1]["added"]) if sorted_by_lines else 1

    print(f"\n  {'Lines':>7} {'':>10}  File")
    print(f"  {'─' * 7} {'─' * 10}  {'─' * 45}")

    for fp, s in sorted_by_lines:
        total_lines = s["written"] + s["added"]
        display = fp if len(fp) <= 50 else "..." + fp[-(47):]
        print(f"  {total_lines:>7} {bar(total_lines, max_lines, 10)}  {display}")

    # ── Insights
    print_header("💡  INSIGHTS", "─")
    insights = []

    if total_written > total_added:
        pct_new = total_written / (total_written + total_added) * 100
        insights.append(f"Mostly new files: {pct_new:.0f}% of lines are from Write (new files/rewrites).")
    else:
        pct_edit = total_added / (total_written + total_added) * 100
        insights.append(f"Mostly edits: {pct_edit:.0f}% of lines are from Edit operations.")

    if total_removed > 0:
        ratio = (total_written + total_added) / total_removed
        insights.append(f"Add/remove ratio: {ratio:.1f}:1 — {'growing codebase' if ratio > 2 else 'healthy refactoring' if ratio > 1 else 'net reduction'}.")

    if sorted_files:
        top_file = sorted_files[0]
        insights.append(f"Most touched file: {top_file[0][-40:]} ({top_file[1]['ops']} operations)")

    if sorted_ext:
        top_ext = sorted_ext[0]
        ext_total = top_ext[1]["written"] + top_ext[1]["added"]
        all_total = total_written + total_added
        insights.append(f"Top language: {top_ext[0]} ({ext_total / all_total * 100:.0f}% of lines)")

    for i, insight in enumerate(insights, 1):
        print(f"  {i}. {insight}")
    print()

File: internal/commands/lines.go
package commands

import (
	"fmt"
	"sort"

	"github.com/Andrevops/claude-stats/internal/dates"
	"github.com/Andrevops/claude-stats/internal/format"
	"github.com/Andrevops/claude-stats/internal/projects"
	"github.com/Andrevops/claude-stats/internal/session"
)

type lineStat struct {
	written, added, removed int
	files                   map[string]bool
}

type fileLineStat struct {
	written, added, removed, ops int
}

func Lines(args []string) {
	targetDates, label := dates.ParseArgs(args)
	if label == "" {
		return
	}
	files := session.Find(targetDates, false)
	if len(files) == 0 {
		fmt.Printf("\n  No sessions found for %s\n", label)
		return
	}

	totalWrites, totalEdits := 0, 0
	totalWritten, totalAdded, totalRemoved := 0, 0, 0

	extStats := map[string]*lineStat{}
	projStats := map[string]*struct {
		lineStat
		writes, edits int
	}{}
	fileStats := map[string]*fileLineStat{}
	allFiles := map[string]bool{}

	for _, f := range files {
		projName := projects.ExtractProject(f)
		session.ScanLines(f, func(line session.LogLine) {
			if line.Type != "assistant" {
				return
			}
			msg, ok := session.ParseAssistantMsg(line.Message)
			if !ok {
				return
			}
			for _, b := range msg.Content {
				if b.Type != "tool_use" {
					continue
				}
				switch b.Name {
				case "Write":
					wi := session.ParseWriteInput(b.Input)
					lines := session.CountLines(wi.Content)
					ext := projects.GetExt(wi.FilePath)
					short := projects.ShortenPath(wi.FilePath)

					totalWrites++
					totalWritten += lines
					allFiles[wi.FilePath] = true

					ensureLineStat(extStats, ext).written += lines
					ensureLineStat(extStats, ext).files[wi.FilePath] = true

					ensureProjStat(projStats, projName).written += lines
					ensureProjStat(projStats, projName).writes++
					ensureProjStat(projStats, projName).files[wi.FilePath] = true

					ensureFileStat(fileStats, short).written += lines
					ensureFileStat(fileStats, short).ops++

				case "Edit":
					ei := session.ParseEditInput(b.Input)
					oldLines := session.CountLines(ei.OldString)
					newLines := session.CountLines(ei.NewString)
					ext := projects.GetExt(ei.FilePath)
					short := projects.ShortenPath(ei.FilePath)

					totalEdits++
					totalAdded += newLines
					totalRemoved += oldLines
					allFiles[ei.FilePath] = true

					ensureLineStat(extStats, ext).added += newLines
					ensureLineStat(extStats, ext).removed += oldLines
					ensureLineStat(extStats, ext).files[ei.FilePath] = true

					ensureProjStat(projStats, projName).added += newLines
					ensureProjStat(projStats, projName).removed += oldLines
					ensureProjStat(projStats, projName).edits++
					ensureProjStat(projStats, projName).files[ei.FilePath] = true

					ensureFileStat(fileStats, short).added += newLines
					ensureFileStat(fileStats, short).removed += oldLines
					ensureFileStat(fileStats, short).ops++
				}
			}
		})
	}

	totalOps := totalWrites + totalEdits
	if totalOps == 0 {
		fmt.Printf("\n  No Write/Edit operations found for %s\n", label)
		return
	}

	net := totalWritten + totalAdded - totalRemoved
	throughput := totalWritten + totalAdded + totalRemoved

	format.Header(fmt.Sprintf("📝  CLAUDE CODE LINES OF CODE — %s", label), "═")
	fmt.Printf(`
  Operations:    %d writes + %d edits = %d total
  Files touched: %d

  ┌──────────────────────────────────────┐
  │  Lines written (new files):  %8s │
  │  Lines added (edits):        %8s │
  │  Lines removed (edits):      %8s │
  │  ──────────────────────────────────  │
  │  NET LINES:                  %8s │
  │  TOTAL THROUGHPUT:           %8s │
  └──────────────────────────────────────┘`+"\n",
		totalWrites, totalEdits, totalOps,
		len(allFiles),
		format.Fmt(totalWritten),
		format.Fmt(totalAdded),
		format.Fmt(totalRemoved),
		format.Fmt(net),
		format.Fmt(throughput),
	)

	// ── By Extension
	format.Header("📄  BY FILE EXTENSION", "─")
	type extEntry struct {
		ext string
		s   *lineStat
	}
	var extList []extEntry
	for ext, s := range extStats {
		extList = append(extList, extEntry{ext, s})
	}
	sort.Slice(extList, func(i, j int) bool {
		ti := extList[i].s.written + extList[i].s.added
		tj := extList[j].s.written + extList[j].s.added
		return ti > tj
	})
	if len(extList) > 15 {
		extList = extList[:15]
	}
	maxExt := 0
	if len(extList) > 0 {
		maxExt = extList[0].s.written + extList[0].s.added
	}

	fmt.Printf("\n  %-10s %8s %8s %8s %8s %5s  %s\n",
		"Ext", "Written", "Added", "Removed", "Net", "Files", "")
	fmt.Printf("  %s %s %s %s %s %s  %s\n",
		repeat("─", 10), repeat("─", 8), repeat("─", 8), repeat("─", 8), repeat("─", 8), repeat("─", 5), repeat("─", 12))

	for _, e := range extList {
		extNet := e.s.written + e.s.added - e.s.removed
		total := e.s.written + e.s.added
		sign := "+"
		if extNet < 0 {
			sign = ""
		}
		fmt.Printf("  %-10s %8d %8d %8d %s%7d %5d  %s\n",
			e.ext, e.s.written, e.s.added, e.s.removed,
			sign, extNet, len(e.s.files),
			format.Bar(float64(total), float64(maxExt), 12))
	}

	// ── By Project
	format.Header("📁  BY PROJECT", "─")
	type projEntry struct {
		name  string
		s     *lineStat
		w, ed int
	}
	var projList []projEntry
	for name, p := range projStats {
		projList = append(projList, projEntry{name, &p.lineStat, p.writes, p.edits})
	}
	sort.Slice(projList, func(i, j int) bool {
		ti := projList[i].s.written + projList[i].s.added
		tj := projList[j].s.written + projList[j].s.added
		return ti > tj
	})
	maxProj := 0
	if len(projList) > 0 {
		maxProj = projList[0].s.written + projList[0].s.added
	}

	fmt.Printf("\n  %-40s %8s %8s %8s %7s  %s\n",
		"Project", "Written", "Added", "Removed", "Net", "")
	fmt.Printf("  %s %s %s %s %s  %s\n",
		repeat("─", 40), repeat("─", 8), repeat("─", 8), repeat("─", 8), repeat("─", 7), repeat("─", 10))

	for _, e := range projList {
		projNet := e.s.written + e.s.added - e.s.removed
		total := e.s.written + e.s.added
		sign := "+"
		if projNet < 0 {
			sign = ""
		}
		name := truncate(e.name, 38)
		fmt.Printf("  %-40s %8d %8d %8d %s%6d  %s\n",
			name, e.s.written, e.s.added, e.s.removed,
			sign, projNet,
			format.Bar(float64(total), float64(maxProj), 10))
	}

	// ── Top Files (most operations)
	format.Header("🔥  TOP FILES (most operations)", "─")
	type fileEntry struct {
		path string
		s    *fileLineStat
	}
	var fileList []fileEntry
	for path, s := range fileStats {
		fileList = append(fileList, fileEntry{path, s})
	}
	sort.Slice(fileList, func(i, j int) bool { return fileList[i].s.ops > fileList[j].s.ops })
	if len(fileList) > 20 {
		fileList = fileList[:20]
	}

	fmt.Printf("\n  %4s %8s %7s %7s  File\n", "Ops", "Written", "+Lines", "-Lines")
	fmt.Printf("  %s %s %s %s  %s\n",
		repeat("─", 4), repeat("─", 8), repeat("─", 7), repeat("─", 7), repeat("─", 45))
	for _, e := range fileList {
		display := e.path
		if len(display) > 50 {
			display = "..." + display[len(display)-47:]
		}
		fmt.Printf("  %4d %8d %7d %7d  %s\n",
			e.s.ops, e.s.written, e.s.added, e.s.removed, display)
	}

	// ── Top Files (most lines)
	format.Header("📏  TOP FILES (most lines written/added)", "─")
	sort.Slice(fileList, func(i, j int) bool {
		ti := fileList[i].s.written + fileList[i].s.added
		tj := fileList[j].s.written + fileList[j].s.added
		return ti > tj
	})
	topFiles := fileList
	if len(topFiles) > 15 {
		topFiles = topFiles[:15]
	}
	maxLines := 0
	if len(topFiles) > 0 {
		maxLines = topFiles[0].s.written + topFiles[0].s.added
	}

	fmt.Printf("\n  %7s %10s  File\n", "Lines", "")
	fmt.Printf("  %s %s  %s\n", repeat("─", 7), repeat("─", 10), repeat("─", 45))
	for _, e := range topFiles {
		total := e.s.written + e.s.added
		display := e.path
		if len(display) > 50 {
			display = "..." + display[len(display)-47:]
		}
		fmt.Printf("  %7d %s  %s\n",
			total, format.Bar(float64(total), float64(maxLines), 10), display)
	}

	// ── Insights
	format.Header("💡  INSIGHTS", "─")
	var insights []string
	if totalWritten > totalAdded {
		pct := float64(totalWritten) / float64(totalWritten+totalAdded) * 100
		insights = append(insights, fmt.Sprintf("Mostly new files: %.0f%% of lines are from Write (new files/rewrites).", pct))
	} else if totalAdded > 0 {
		pct := float64(totalAdded) / float64(totalWritten+totalAdded) * 100
		insights = append(insights, fmt.Sprintf("Mostly edits: %.0f%% of lines are from Edit operations.", pct))
	}
	if totalRemoved > 0 {
		ratio := float64(totalWritten+totalAdded) / float64(totalRemoved)
		desc := "growing codebase"
		if ratio <= 2 {
			desc = "healthy refactoring"
		}
		if ratio <= 1 {
			desc = "net reduction"
		}
		insights = append(insights, fmt.Sprintf("Add/remove ratio: %.1f:1 — %s.", ratio, desc))
	}
	if len(fileList) > 0 {
		tf := fileList[0]
		insights = append(insights, fmt.Sprintf("Most touched file: %s (%d operations)", truncate(tf.path, 40), tf.s.ops))
	}
	if len(extList) > 0 {
		te := extList[0]
		te_total := te.s.written + te.s.added
		all_total := totalWritten + totalAdded
		if all_total > 0 {
			insights = append(insights, fmt.Sprintf("Top language: %s (%.0f%% of lines)", te.ext, float64(te_total)/float64(all_total)*100))
		}
	}

	for i, ins := range insights {
		fmt.Printf("  %d. %s\n", i+1, ins)
	}
	fmt.Println()
}

func ensureLineStat(m map[string]*lineStat, key string) *lineStat {
	if _, ok := m[key]; !ok {
		m[key] = &lineStat{files: map[string]bool{}}
	}
	return m[key]
}

func ensureProjStat(m map[string]*struct {
	lineStat
	writes, edits int
}, key string) *struct {
	lineStat
	writes, edits int
} {
	if _, ok := m[key]; !ok {
		m[key] = &struct {
			lineStat
			writes, edits int
		}{lineStat: lineStat{files: map[string]bool{}}}
	}
	return m[key]
}

func ensureFileStat(m map[string]*fileLineStat, key string) *fileLineStat {
	if _, ok := m[key]; !ok {
		m[key] = &fileLineStat{}
	}
	return m[key]
}
